### PR TITLE
Detect the managed instance group that created a VM

### PR DESCRIPTION
**Pull request by quentinmit**
In preparation for new resource attributes for the MIG.

### CODE REVIEW

File: detectors/gcp/gce.go [hunk 1]
```diff
@@ -73,3 +79,36 @@ func (d *Detector) GCEAvailabilityZoneAndRegion() (string, string, error) {
...
+func (d *Detector) GCEManagedInstanceGroup() (ManagedInstanceGroup, error) {
+	createdBy, err := d.metadata.Get(createdByMetadataAttr)
+	if _, ok := err.(metadata.NotDefinedError); ok {
+		return ManagedInstanceGroup{}, nil
+	} else if err != nil {
+		return ManagedInstanceGroup{}, err
+	}
+	matches := createdByMIGRE.FindStringSubmatch(createdBy)
+	if matches == nil {
+		return ManagedInstanceGroup{}, nil
```
Comment (dashpole) [inline detectors/gcp/gce.go:100]: Should we return an error here? Seems like if it is in the wrong format we should let the user know, right?

Comment (quentinmit) [reply detectors/gcp/gce.go:100]: The key is just `created-by`; my assumption is that other tools will put their own strings indicating that they created the VM, and that's not an error, it just means the VM wasn't created by a MIG.

Comment (dashpole) [reply detectors/gcp/gce.go:100]: please add a comment :)

File: detectors/gcp/utils_test.go [hunk 2]
```diff
@@ -14,62 +14,87 @@
...
+	fakeProjectNumber       = "1234567890"
+	fakeInstanceID          = "1234567891"
+	fakeInstanceName        = "my-instance"
+	fakeInstanceHostname    = fakeInstanceName + "." + fakeZone + ".c." + fakeProjectID + ".internal"
+	fakeInstanceMachineType = "n1-standard1"
+)
+
+// fmp constructs a FakeMetadatProvider that responds with the basic GCE metadata fields.
+// To add more metadata, pass key-value pairs as arguments.
+func fmp(keyValues ...string) *FakeMetadataProvider {
```
Comment (dashpole) [inline detectors/gcp/utils_test.go:82]: nit: i'd prefer newFakeMetadataProvider to the abbreviation for readability.

File: detectors/gcp/utils_test.go [hunk 3]
```diff
@@ -14,62 +14,87 @@
...
 }
 func (f *FakeMetadataProvider) Zone() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	value, err := f.Get("instance/zone")
+	if err != nil {
+		return value, err
 	}
-	return f.FakeZone, nil
+	// https://github.com/googleapis/google-cloud-go/blob/35d1f813f755c88d6869d993796c5ddf8ac3eea3/compute/metadata/metadata.go#L663
```
Comment (dashpole) [inline detectors/gcp/utils_test.go:62]: This link seems irrelevant?

File: detectors/gcp/utils_test.go [hunk 4]
```diff
@@ -14,62 +14,87 @@
...
+	return value[strings.LastIndex(value, "/")+1:], nil
```
Comment (dashpole) [inline detectors/gcp/utils_test.go:63]: I'm not sure this is helpful for testing our library, even if it is closer to how the metadata library works.

File: detectors/gcp/gce.go [hunk 5]
```diff
@@ -73,3 +79,38 @@ func (d *Detector) GCEAvailabilityZoneAndRegion() (string, string, error) {
 	}
 	return zone, strings.Join(splitZone[0:2], "-"), nil
 }
+
+type ManagedInstanceGroup struct {
+	Name     string
+	Location string
```
Comment (dashpole) [inline detectors/gcp/gce.go:85]: Do you think we should omit the location + location type until we are sure we actually need the location attributes in the conventions?

Comment (quentinmit) [reply detectors/gcp/gce.go:85]: We definitely need a location attribute; there's otherwise no way to distinguish a zonal MIG named `foo` and a regional MIG named `foo`. I guess it's an open question how exactly that information is encoded into attributes, but we do need it in some form.

File: detectors/gcp/utils_test.go (hunk 4 above)
Comment (quentinmit) [reply detectors/gcp/utils_test.go:63]: Yeah, I was torn on whether to do this or not, but I came down on the side of wanting `Get("instance/zone")` and `Zone()` to always return matching values, which means copying this logic.

Now that I've done this simplification we could also get rid of `FakeMetadatProvider` entirely if we want and just use the upstream `metadata.Client` with a mock HTTP handler. Then we'd use upstream's split logic here. WDYT?

Comment (dashpole) [reply detectors/gcp/utils_test.go:63]: sounds good.  lets do it

File: detectors/gcp/gce.go (hunk 5 above)
Comment (dashpole) [reply detectors/gcp/gce.go:85]: My question is: is the location of the instance (not the MIG) enough, or do we need both?  The location of the instance is already provided via GCEAvailabilityZoneAndRegion

File: detectors/gcp/utils_test.go (hunk 2 above)
Comment (quentinmit) [reply detectors/gcp/utils_test.go]: Done

File: detectors/gcp/gce.go (hunk 5 above)
Comment (quentinmit) [reply detectors/gcp/gce.go:86]: The location of the instance is always a zone. The location of the MIG can be a zone or region, so no, the location of the instance is not enough by itself. I *think* it might be the case that the location of the MIG is always *either* the zone or the region of the instance, so it could be modeled as *either* `mig.type: regional` (using only what I've called `Type` here) or `mig.region: us-central1` (using both `Location` and `Type`). I think OTel generally doesn't have `type` attributes like that, so I lean towards the latter.

But I can't find any documentation that actually asserts that MIGs will always be in the same region or zone as the instance, so I'm hedging my bets on the API here and including both.

Comment (dashpole) [reply detectors/gcp/gce.go:86]: ok.  worst case, we have extra API surface we don't use

**Review by dashpole — APPROVED** [review]

**Comment by dashpole** [comment]
Code changes LGTM.  Please document the steps you took to manually test this PR in the description before merging.
